--- ./crypto/cmac/cmac.c	2019-09-04 14:50:24.034778060 +0900
+++ ../talos/./crypto/cmac/cmac.c	2019-09-04 14:49:11.878299784 +0900
@@ -107,10 +107,10 @@
 CMAC_CTX_cleanup(CMAC_CTX *ctx)
 {
 	EVP_CIPHER_CTX_cleanup(&ctx->cctx);
-	explicit_bzero(ctx->tbl, EVP_MAX_BLOCK_LENGTH);
-	explicit_bzero(ctx->k1, EVP_MAX_BLOCK_LENGTH);
-	explicit_bzero(ctx->k2, EVP_MAX_BLOCK_LENGTH);
-	explicit_bzero(ctx->last_block, EVP_MAX_BLOCK_LENGTH);
+	bzero(ctx->tbl, EVP_MAX_BLOCK_LENGTH);
+	bzero(ctx->k1, EVP_MAX_BLOCK_LENGTH);
+	bzero(ctx->k2, EVP_MAX_BLOCK_LENGTH);
+	bzero(ctx->last_block, EVP_MAX_BLOCK_LENGTH);
 	ctx->nlast_block = -1;
 }
 
@@ -183,7 +183,7 @@
 			return 0;
 		make_kn(ctx->k1, ctx->tbl, bl);
 		make_kn(ctx->k2, ctx->k1, bl);
-		explicit_bzero(ctx->tbl, bl);
+		bzero(ctx->tbl, bl);
 		/* Reset context again ready for first data block */
 		if (!EVP_EncryptInit_ex(&ctx->cctx, NULL, NULL, NULL, zero_iv))
 			return 0;
@@ -260,7 +260,7 @@
 			out[i] = ctx->last_block[i] ^ ctx->k2[i];
 	}
 	if (!EVP_Cipher(&ctx->cctx, out, out, bl)) {
-		explicit_bzero(out, bl);
+		bzero(out, bl);
 		return 0;
 	}
 	return 1;
