--- ./include/openssl/conf.h	2017-03-08 11:15:50.745464080 +0000
+++ ../talos/./include/openssl/conf.h	2017-03-08 11:13:57.483751762 +0000
@@ -134,6 +134,7 @@
 int CONF_dump_fp(LHASH_OF(CONF_VALUE) *conf, FILE *out);
 int CONF_dump_bio(LHASH_OF(CONF_VALUE) *conf, BIO *out);
 
+void ecall_OPENSSL_config(const char *config_name);
 void OPENSSL_config(const char *config_name);
 void OPENSSL_no_config(void);
 
@@ -170,6 +171,7 @@
     unsigned long flags);
 int CONF_modules_load_file(const char *filename, const char *appname,
     unsigned long flags);
+void ecall_CONF_modules_unload(int all);
 void CONF_modules_unload(int all);
 void CONF_modules_finish(void);
 void CONF_modules_free(void);
@@ -191,6 +193,7 @@
 int CONF_parse_list(const char *list, int sep, int nospc,
     int (*list_cb)(const char *elem, int len, void *usr), void *arg);
 
+void ecall_OPENSSL_load_builtin_modules(void);
 void OPENSSL_load_builtin_modules(void);
 
 /* BEGIN ERROR CODES */
