--- ./crypto/dso/dso_dlfcn.c	2017-11-28 10:46:51.996404519 +0000
+++ ../talos/./crypto/dso/dso_dlfcn.c	2017-07-17 11:08:32.213410215 +0100
@@ -100,6 +100,8 @@
 	.globallookup = dlfcn_globallookup
 };
 
+extern int my_asprintf(char **strp, const char *fmt, ...);
+
 DSO_METHOD *
 DSO_METHOD_dlfcn(void)
 {
@@ -295,9 +297,9 @@
 	if (strchr(filename, '/') == NULL) {
 		/* Bare name, so convert to "%s.so" or "lib%s.so" */
 		if ((DSO_flags(dso) & DSO_FLAG_NAME_TRANSLATION_EXT_ONLY) == 0)
-			ret = asprintf(&translated, "lib%s" DSO_ext, filename);
+			ret = my_asprintf(&translated, "lib%s" DSO_ext, filename);
 		else
-			ret = asprintf(&translated, "%s" DSO_ext, filename);
+			ret = my_asprintf(&translated, "%s" DSO_ext, filename);
 		if (ret == -1)
 			translated = NULL;
 	} else {
