--- ./crypto/ec/ec_key.c	2019-09-04 14:50:24.042778113 +0900
+++ ../talos/./crypto/ec/ec_key.c	2019-09-04 14:49:11.922300076 +0900
@@ -91,6 +91,11 @@
 }
 
 EC_KEY *
+ecall_EC_KEY_new_by_curve_name(int nid) {
+	return EC_KEY_new_by_curve_name(nid);
+}
+
+EC_KEY *
 EC_KEY_new_by_curve_name(int nid)
 {
 	EC_KEY *ret = EC_KEY_new();
@@ -105,6 +110,11 @@
 }
 
 void 
+ecall_EC_KEY_free(EC_KEY * r) {
+	EC_KEY_free(r);
+}
+
+void 
 EC_KEY_free(EC_KEY * r)
 {
 	int i;
@@ -122,7 +132,7 @@
 
 	EC_EX_DATA_free_all_data(&r->method_data);
 
-	explicit_bzero((void *) r, sizeof(EC_KEY));
+	bzero((void *) r, sizeof(EC_KEY));
 
 	free(r);
 }
