--- ./crypto/compat/getentropy_solaris.c	2019-09-04 14:50:24.038778087 +0900
+++ ../talos/./crypto/compat/getentropy_solaris.c	2019-09-04 14:49:11.902299944 +0900
@@ -434,8 +434,8 @@
 		memcpy((char *)buf + i, results, min(sizeof(results), len - i));
 		i += min(sizeof(results), len - i);
 	}
-	explicit_bzero(&ctx, sizeof ctx);
-	explicit_bzero(results, sizeof results);
+	bzero(&ctx, sizeof ctx);
+	bzero(results, sizeof results);
 	if (gotdata(buf, len) == 0) {
 		errno = save_errno;
 		return (0);		/* satisfied */
