--- ./crypto/hmac/hm_pmeth.c	2017-11-28 10:46:51.996404519 +0000
+++ ../talos/./crypto/hmac/hm_pmeth.c	2017-07-17 11:08:32.241409681 +0100
@@ -66,6 +66,13 @@
 
 #include "evp_locl.h"
 
+#ifdef COMPILE_WITH_INTEL_SGX
+#include "sgx_error.h"
+#include "sgx_trts.h"
+
+extern sgx_status_t ocall_free(void* ptr);
+#endif
+
 /* HMAC pkey context structure */
 
 typedef struct {
@@ -123,9 +130,18 @@
 
 	HMAC_CTX_cleanup(&hctx->ctx);
 	if (hctx->ktmp.data) {
-		if (hctx->ktmp.length)
-			explicit_bzero(hctx->ktmp.data, hctx->ktmp.length);
-		free(hctx->ktmp.data);
+	    if (hctx->ktmp.length) {
+                explicit_bzero(hctx->ktmp.data, hctx->ktmp.length);
+        }
+#ifdef COMPILE_WITH_INTEL_SGX
+        if (sgx_is_within_enclave(hctx->ktmp.data, 1)) {
+#endif
+        free(hctx->ktmp.data);
+#ifdef COMPILE_WITH_INTEL_SGX
+        } else {
+		ocall_free(hctx->ktmp.data);
+	}
+#endif
 		hctx->ktmp.data = NULL;
 	}
 	free(hctx);
