--- ./crypto/malloc-wrapper.c	2017-11-28 10:46:51.996404519 +0000
+++ ../talos/./crypto/malloc-wrapper.c	2017-07-17 11:08:32.257409376 +0100
@@ -18,6 +18,10 @@
 #include <stdlib.h>
 #include <string.h>
 
+#ifdef COMPILE_WITH_INTEL_SGX
+extern char* my_strdup(const char*);
+#endif
+
 int
 CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t),
     void (*f)(void *))
@@ -140,7 +144,11 @@
 char *
 CRYPTO_strdup(const char *str, const char *file, int line)
 {
+#ifdef COMPILE_WITH_INTEL_SGX
+	return my_strdup(str);
+#else
 	return strdup(str);
+#endif
 }
 
 void *
