--- ./crypto/md5/md5_dgst.c	2017-11-28 10:46:51.996404519 +0000
+++ ../talos/./crypto/md5/md5_dgst.c	2017-07-17 11:08:32.241409681 +0100
@@ -69,15 +69,26 @@
 #define INIT_DATA_C (unsigned long)0x98badcfeL
 #define INIT_DATA_D (unsigned long)0x10325476L
 
+int ecall_MD5_Init(MD5_CTX *c) {
+	return MD5_Init(c);
+}
 int MD5_Init(MD5_CTX *c)
-	{
+{
 	memset (c,0,sizeof(*c));
 	c->A=INIT_DATA_A;
 	c->B=INIT_DATA_B;
 	c->C=INIT_DATA_C;
 	c->D=INIT_DATA_D;
 	return 1;
-	}
+}
+
+int ecall_MD5_Update(MD5_CTX *c, const void *data, size_t len) {
+		 return MD5_Update(c, data, len);
+}
+
+int ecall_MD5_Final(unsigned char *md, MD5_CTX *c) {
+		 return MD5_Final(md, c);
+}
 
 #ifndef md5_block_data_order
 #ifdef X
