--- ./crypto/evp/names.c	2017-11-28 10:46:51.996404519 +0000
+++ ../talos/./crypto/evp/names.c	2017-07-17 11:08:32.237409759 +0100
@@ -121,6 +121,10 @@
 	return (cp);
 }
 
+EVP_MD *ecall_EVP_get_digestbyname(const char *name) {
+	EVP_MD* md = (EVP_MD*)EVP_get_digestbyname(name);
+	return md;
+}
 const EVP_MD *
 EVP_get_digestbyname(const char *name)
 {
@@ -131,6 +135,11 @@
 }
 
 void
+ecall_EVP_cleanup(void)
+{
+	EVP_cleanup();
+}
+void
 EVP_cleanup(void)
 {
 	OBJ_NAME_cleanup(OBJ_NAME_TYPE_CIPHER_METH);
