--- ./crypto/pkcs12/p12_decr.c	2019-09-04 14:50:24.058778219 +0900
+++ ../talos/./crypto/pkcs12/p12_decr.c	2019-09-04 14:49:12.106301297 +0900
@@ -138,7 +138,7 @@
 	p = out;
 	ret = ASN1_item_d2i(NULL, &p, outlen, it);
 	if (zbuf)
-		explicit_bzero(out, outlen);
+		bzero(out, outlen);
 	if (!ret)
 		PKCS12err(PKCS12_F_PKCS12_ITEM_DECRYPT_D2I,
 		    PKCS12_R_DECODE_ERROR);
@@ -177,7 +177,7 @@
 		goto err;
 	}
 	if (zbuf)
-		explicit_bzero(in, inlen);
+		bzero(in, inlen);
 	free(in);
 	return oct;
 
