--- ./crypto/pem/pem_pkey.c	2019-09-04 14:50:24.058778219 +0900
+++ ../talos/./crypto/pem/pem_pkey.c	2019-09-04 14:49:12.090301191 +0900
@@ -75,8 +75,33 @@
 
 #include "asn1_locl.h"
 
+#ifdef COMPILE_WITH_INTEL_SGX
+#include "enclaveshim_ocalls.h"
+#endif
+
 int pem_check_suffix(const char *pem_str, const char *suffix);
 
+#ifdef COMPILE_WITH_INTEL_SGX
+pem_password_cb* pem_read_bio_privatekey_cb_address = NULL;
+
+int pem_read_bio_privatekey_fake_cb(char *buf, int size, int rwflag, void *userdata) {
+	if (pem_read_bio_privatekey_cb_address) {
+		my_fprintf(0, "need to call pem_read_bio_privatekey_cb_address\n");
+	}
+	return 0;
+}
+#endif
+
+EVP_PKEY* ecall_PEM_read_bio_PrivateKey(BIO *bp, EVP_PKEY **x, void* cb, void *u) {
+#ifdef COMPILE_WITH_INTEL_SGX
+	pem_read_bio_privatekey_cb_address = (pem_password_cb*)cb;
+	pem_password_cb* callback = pem_read_bio_privatekey_fake_cb;
+#else
+	pem_password_cb* callback = (pem_password_cb*)cb;
+#endif
+	return PEM_read_bio_PrivateKey(bp, x, callback, u);
+}
+
 EVP_PKEY *
 PEM_read_bio_PrivateKey(BIO *bp, EVP_PKEY **x, pem_password_cb *cb, void *u)
 {
@@ -144,7 +169,7 @@
 		PEMerr(PEM_F_PEM_READ_BIO_PRIVATEKEY, ERR_R_ASN1_LIB);
 err:
 	free(nm);
-	explicit_bzero(data, len);
+	bzero(data, len);
 	free(data);
 	return (ret);
 }
