--- ./crypto/pkcs7/pk7_doit.c	2019-09-04 14:50:24.058778219 +0900
+++ ../talos/./crypto/pkcs7/pk7_doit.c	2019-09-04 14:49:12.114301350 +0900
@@ -233,7 +233,7 @@
 	ret = 1;
 
 	if (*pek) {
-		explicit_bzero(*pek, *peklen);
+		bzero(*pek, *peklen);
 		free(*pek);
 	}
 
@@ -371,7 +371,7 @@
 			if (pkcs7_encode_rinfo(ri, key, keylen) <= 0)
 				goto err;
 		}
-		explicit_bzero(key, keylen);
+		bzero(key, keylen);
 
 		if (out == NULL)
 			out = btmp;
@@ -588,7 +588,7 @@
 			 */
 			if (!EVP_CIPHER_CTX_set_key_length(evp_ctx, eklen)) {
 				/* Use random key as MMA defence */
-				explicit_bzero(ek, eklen);
+				bzero(ek, eklen);
 				free(ek);
 				ek = tkey;
 				eklen = tkeylen;
@@ -601,12 +601,12 @@
 			goto err;
 
 		if (ek) {
-			explicit_bzero(ek, eklen);
+			bzero(ek, eklen);
 			free(ek);
 			ek = NULL;
 		}
 		if (tkey) {
-			explicit_bzero(tkey, tkeylen);
+			bzero(tkey, tkeylen);
 			free(tkey);
 			tkey = NULL;
 		}
@@ -635,11 +635,11 @@
 	if (0) {
 err:
 		if (ek) {
-			explicit_bzero(ek, eklen);
+			bzero(ek, eklen);
 			free(ek);
 		}
 		if (tkey) {
-			explicit_bzero(tkey, tkeylen);
+			bzero(tkey, tkeylen);
 			free(tkey);
 		}
 		if (out != NULL)
