--- ./crypto/sha/sha512.c	2019-09-04 14:50:24.058778219 +0900
+++ ../talos/./crypto/sha/sha512.c	2019-09-04 14:49:12.138301509 +0900
@@ -248,7 +248,7 @@
 	SHA384_Init(&c);
 	SHA512_Update(&c,d,n);
 	SHA512_Final(md,&c);
-	explicit_bzero(&c,sizeof(c));
+	bzero(&c,sizeof(c));
 	return(md);
 	}
 
@@ -261,7 +261,7 @@
 	SHA512_Init(&c);
 	SHA512_Update(&c,d,n);
 	SHA512_Final(md,&c);
-	explicit_bzero(&c,sizeof(c));
+	bzero(&c,sizeof(c));
 	return(md);
 	}
 
