--- ./ssl/ssl_stat.c	2017-11-28 10:46:52.004404366 +0000
+++ ../talos/./ssl/ssl_stat.c	2017-07-17 11:08:26.857512440 +0100
@@ -86,6 +86,30 @@
 
 #include "ssl_locl.h"
 
+#ifdef COMPILE_WITH_INTEL_SGX
+#include "sgx_error.h"
+
+extern sgx_status_t ocall_malloc(void** retval, size_t size);
+
+char* SSL_state_string_long_buffer = NULL;
+#endif
+
+char*
+ecall_SSL_state_string_long(const SSL *s)
+{
+	char* ret = (char*) SSL_state_string_long(s);
+#ifdef COMPILE_WITH_INTEL_SGX
+	if (!SSL_state_string_long_buffer) {
+		ocall_malloc((void*)&SSL_state_string_long_buffer, 8192);
+	}
+	size_t sz = strlen(ret);
+	memcpy(SSL_state_string_long_buffer, ret, sz);
+	SSL_state_string_long_buffer[sz] = '\0';
+	ret = SSL_state_string_long_buffer;
+#endif
+	return ret;
+}
+
 const char *
 SSL_state_string_long(const SSL *s)
 {
