--- ./crypto/x509/x_all.c	2017-03-08 11:15:50.729464404 +0000
+++ ../talos/./crypto/x509/x_all.c	2017-03-08 11:13:57.455752326 +0000
@@ -496,6 +496,10 @@
 }
 
 int
+ecall_X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md) {
+	return X509_sign(x, pkey, md);
+}
+int
 X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md)
 {
 	x->cert_info->enc.modified = 1;
@@ -551,6 +555,11 @@
 }
 
 int
+ecall_X509_pubkey_digest(const X509 *data, const EVP_MD *type, unsigned char *md,
+    unsigned int *len) {
+	return X509_pubkey_digest(data, type, md, len);
+}
+int
 X509_pubkey_digest(const X509 *data, const EVP_MD *type, unsigned char *md,
     unsigned int *len)
 {
@@ -562,6 +571,13 @@
 }
 
 int
+ecall_X509_digest(const X509 *data, const EVP_MD *type, unsigned char *md,
+    unsigned int *len)
+{
+	return X509_digest(data, type, md, len);
+}
+
+int
 X509_digest(const X509 *data, const EVP_MD *type, unsigned char *md,
     unsigned int *len)
 {
